js_modules/dagit/packages/core/src/app/Permissions.tsx

```diff
@@ -55,60 +58,122 @@ const extractPermissions = (permissions: PermissionFragmentFragment[]) => {
+    canReloadWorkspace: permissionOrFallback('reload_workspace'),
+    canWipeAssets: permissionOrFallback('wipe_assets'),
+    canLaunchPartitionBackfill: permissionOrFallback('launch_partition_backfill'),
+    canCancelPartitionBackfill: permissionOrFallback('cancel_partition_backfill'),
   };
 };
 
 export type PermissionsMap = ReturnType<typeof extractPermissions>;
 
 export const PermissionsContext = React.createContext<{
-  data: PermissionFragmentFragment[];
+  fallbackPermissions: PermissionFragmentFragment[];
+  locationPermissions: Record<string, PermissionFragmentFragment[]>;
   loading: boolean;
-}>({data: [], loading: true});
+}>({fallbackPermissions: [], locationPermissions: {}, loading: true});
 
 export const PermissionsProvider: React.FC = (props) => {
   const {data, loading} = useQuery(PERMISSIONS_QUERY, {
     fetchPolicy: 'cache-first', // Not expected to change after initial load.
   });
-  const value = React.useMemo(
-    () => ({
-      data: data?.permissions || [],
+
+  const value = React.useMemo(() => {
+    const locationEntries =
+      data?.workspaceOrError.__typename === 'Workspace'
+        ? data.workspaceOrError.locationEntries
+        : [];
+
+    const locationMap = {};
+    locationEntries.forEach((locationEntry) => {
+      const {name, permissions} = locationEntry;
+      locationMap[name] = permissions;
+    });
+
+    return {
+      fallbackPermissions: data?.unscopedPermissions || [],
+      locationPermissions: locationMap,
       loading,
-    }),
-    [data, loading],
-  );
+    };
+  }, [data, loading]);
   return <PermissionsContext.Provider value={value}>{props.children}</PermissionsContext.Provider>;
 };
 
-export const usePermissions = () => {
-  const {data, loading} = React.useContext(PermissionsContext);
-  return React.useMemo(
-    () => ({
-      ...extractPermissions(data),
+/**
+ * Retrieve a permission that is intentionally unscoped.
+ */
+export const useUnscopedPermissions = () => {
+  const {fallbackPermissions, loading} = React.useContext(PermissionsContext);
+  return React.useMemo(() => {
+    return {
+      ...extractPermissions(fallbackPermissions),
       loading,
-    }),
-    [data, loading],
-  );
+    };
+  }, [fallbackPermissions, loading]);
 };
 
+/**
+ * Retrieve a permission that is scoped to a specific code location. The unscoped permission set
+ * will be used as a fallback, so that if the permission is not defined for that location, we still
+ * have a valid value.
+ */
+export const usePermissionsForLocation = (locationName: string | null | undefined) => {
+  const {fallbackPermissions, locationPermissions, loading} = React.useContext(PermissionsContext);
+  return React.useMemo(() => {
+    const targetPermissions = locationName
+      ? locationPermissions[locationName]
+      : fallbackPermissions;
+    return {
+      ...extractPermissions(targetPermissions || [], fallbackPermissions),
```

Comment:
This all looks good! It looks like until the data has loaded, the user has no permissions which seems good. That said I wonder if there's any chance the cache could include fallback permissions but an old version of the location data, resulting in a user briefly getting fallback permissions for a new location instead of the location permissions?

Sounds like it won't matter once the migration is complete and the hook usage is updated?